pkg/comp-functions/functions/common/billing_service.go

```diff
@@ -0,0 +1,196 @@
+		log.Info("Test instance, skipping billing")
+		return runtime.NewNormalResult(fmt.Sprintf("billing skipped for test instance %s", comp.GetName()))
+	}
+
+	claim := comp.GetClaimName()
+	namespace := comp.GetClaimNamespace()
+	service := comp.GetServiceName()
+
+	// Form productID from service and number of replicas (or use override)
+	productID := opts.ProductID
+	if productID == "" {
+		productID = getProductID(comp.GetInstances(), service)
+	}
+
+	// Get unitID from config
+	unitID := svc.Config.Data["billingUnitID"]
+	if unitID == "" {
+		log.Error(fmt.Errorf("missing billing unitID"), "UnitID missing in composition")
+		return runtime.NewWarningResult(fmt.Sprintf("no billing unit id set in composition for %s", comp.GetName()))
+	}
+
+	// Get clusterName from config
+	clusterName := svc.Config.Data["clusterName"]
+	if clusterName == "" {
+		log.Error(fmt.Errorf("missing billing clusterName"), "clusterName missing in composition")
+		return runtime.NewWarningResult(fmt.Sprintf("no clusterName set in composition for %s", comp.GetName()))
+	}
+
+	// Get keepAfterDeletion from config
+	keepAfterDeletion := DefaultKeepAfterDeletion
+	if keepAfterDeletionStr := svc.Config.Data["crDeletionAfter"]; keepAfterDeletionStr != "" {
+		if val, err := strconv.Atoi(keepAfterDeletionStr); err == nil {
+			keepAfterDeletion = val
+		}
+	}
+
+	isAPPUiOCloud := false
+	salesOrder := svc.Config.Data["salesOrder"]
+	if salesOrder == "" {
+		isAPPUiOCloud = true
+	}
+
+	// Determine size (use override or instance count)
+	size := opts.Size
+	if size == "" {
+		size = strconv.Itoa(comp.GetInstances())
+	}
+
+	// Build labels
+	labels := map[string]string{
+		"appcat.vshn.io/claim-name":      claim,
+		"appcat.vshn.io/claim-namespace": namespace,
+		"appcat.vshn.io/service-name":    service,
+	}
+	for k, v := range opts.AdditionalLabels {
+		labels[k] = v
+	}
+
+	// Create BillingService CR
+	billingService := &vshnv1.BillingService{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      comp.GetName() + opts.ResourceNameSuffix,
+			Namespace: BillingNamespace,
+			Labels:    labels,
+		},
+		Spec: vshnv1.BillingServiceSpec{
+			KeepAfterDeletion: keepAfterDeletion,
+			Odoo: vshnv1.OdooSpec{
+				InstanceID:           comp.GetName(),
+				ProductID:            productID,
+				UnitID:               unitID,
+				Size:                 size,
+				SalesOrderID:         salesOrder,
+				ItemGroupDescription: claim,
+				ItemDescription:      GetItemDescription(isAPPUiOCloud, clusterName, namespace),
+			},
+		},
+	}
+
+	// Get organization (APPUiO cloud)
+	if isAPPUiOCloud {
+		org, err := GetOrg(comp.GetName(), svc)
+		if err != nil {
+			log.Error(err, "billing sales order and organization are missing", "service", comp.GetName())
+			return runtime.NewWarningResult(fmt.Sprintf("cannot add billing to service %s", comp.GetName()))
+		}
+		billingService.Spec.Odoo.Organization = org
+	}
+
+	kubeObj := &xkube.Object{}
+	ownerRefOption := func(obj *xkube.Object) {}
```

Comment:
I'd probably declare a `[]KubeOPtion` and conditionally add the option for the ownerref in the if below.

Then you don't have to ad-hoc declare this ugly noop...